File: go/leetcode/15.三数之和.go
/*
 * @lc app=leetcode.cn id=15 lang=golang
 *
 * [15] 三数之和
 *
 * https://leetcode-cn.com/problems/3sum/description/
 *
 * algorithms
 * Medium (22.51%)
 * Likes:    1015
 * Dislikes: 0
 * Total Accepted:    58.1K
 * Total Submissions: 258.1K
 * Testcase Example:  '[-1,0,1,2,-1,-4]'
 *
 * 给定一个包含 n 个整数的数组 nums，判断 nums 中是否存在三个元素 a，b，c ，使得 a + b + c = 0
 * ？找出所有满足条件且不重复的三元组。
 * 
 * 注意：答案中不可以包含重复的三元组。
 * 
 * 例如, 给定数组 nums = [-1, 0, 1, 2, -1, -4]，
 * 
 * 满足要求的三元组集合为：
 * [
 * ⁠ [-1, 0, 1],
 * ⁠ [-1, -1, 2]
 * ]
 * 
 * 
 */

import (
	"sort"
)

// 1. 先排序
// 2. 固定一个，后两个数从边缘往中间搜索
// 3. 做一些减枝优化
func threeSum(nums []int) [][]int {
	sort.Ints(nums)
	ret := [][]int{}
	// 三元组，所以只需要遍历到倒数第三个
	for i := 0; i < len(nums) - 2; i++ {
		// 剪枝优化1，最小值>0则退出
		if nums[i] > 0 {
			break
		}
		// 去重1，与前面相同则不再搜索
		if i > 0 && nums[i] == nums[i-1] {
			continue
		}
		// 缓存target
		target := 0 - nums[i]
		// 双指针分别指向两侧边缘
		j, k := i+1, len(nums)-1
		for j < k {
			if nums[j] + nums[k] == target {
				ret = append(ret, []int{nums[i], nums[j], nums[k]})
				// 去重2
				for j < k && nums[j] == nums[j + 1] {
					j++
				}
				// 去重3
				for j < k && nums[k] == nums[k - 1] {
					k--
				}
				j++
				k--
			} else if nums[j] + nums[k] < target {
				j++
			} else {
				k--
			}
		}
	}
	return ret
}

File: cpp/leetcode/15.三数之和.cpp
/*
 * @lc app=leetcode.cn id=15 lang=cpp
 *
 * [15] 三数之和
 *
 * https://leetcode-cn.com/problems/3sum/description/
 *
 * algorithms
 * Medium (32.32%)
 * Likes:    3410
 * Dislikes: 0
 * Total Accepted:    532.8K
 * Total Submissions: 1.6M
 * Testcase Example:  '[-1,0,1,2,-1,-4]'
 *
 * 给你一个包含 n 个整数的数组 nums，判断 nums 中是否存在三个元素 a，b，c
 * ，使得 a + b + c = 0 ？请你找出所有和为 0 且不重复的三元组。
 *
 * 注意：答案中不可以包含重复的三元组。
 *
 *
 *
 * 示例 1：
 *
 *
 * 输入：nums = [-1,0,1,2,-1,-4]
 * 输出：[[-1,-1,2],[-1,0,1]]
 *
 *
 * 示例 2：
 *
 *
 * 输入：nums = []
 * 输出：[]
 *
 *
 * 示例 3：
 *
 *
 * 输入：nums = [0]
 * 输出：[]
 *
 *
 *
 *
 * 提示：
 *
 *
 * 0
 * -10^5
 *
 *
 */

// @lc code=start
#include <algorithm>
#include <iostream>
#include <ostream>
#include <vector>
class Solution {
public:
  std::vector<std::vector<int>> threeSum(std::vector<int> &nums) {
    std::vector<std::vector<int>> ret;
    if (nums.size() < 3) {
      return ret;
    }
    std::sort(nums.begin(), nums.end());
    if (nums.at(nums.size() - 1) < 0) {
      return ret;
    }
    for (size_t i = 0; i < nums.size() - 2; i++) {
      if (nums.at(0) > 0) {
        break;
      }
      if (i > 0 && nums.at(i) == nums.at(i - 1)) {
        continue;
      }
      int rest = 0 - nums.at(i);
      int l = i + 1, r = nums.size() - 1;
      while (l < r) {
        int tmp = nums.at(l) + nums.at(r);
        if (tmp == rest) {
          ret.push_back({nums.at(i), nums.at(l), nums.at(r)});
          while (l < r && nums.at(l) == nums.at(l + 1)) {
            l++;
          }
          while (l < r && nums.at(r) == nums.at(r - 1)) {
            r--;
          }
          l++;
          r--;
        } else if (tmp < rest) {
          l++;
        } else {
          r--;
        }
      }
    }
    return ret;
  }
};
// @lc code=end

void print_ret(const std::vector<std::vector<int>> &ret) {
  std::cout << "ret: [" << std::endl;
  for (size_t i = 0; i < ret.size(); i++) {
    std::cout << "[";
    for (size_t j = 0; j < ret.at(i).size(); j++) {
      std::cout << ret.at(i).at(j) << ",";
    }
    std::cout << "]" << std::endl;
  }
  std::cout << "]" << std::endl;
}

int main(int argc, const char **argv) {
  Solution s;
  std::vector<int> arr = {-1, 0, 1, 2, -1, -4};
  auto ret = s.threeSum(arr);
  print_ret(ret);
  return 0;
}